apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/S3DirectoryDao.java

```diff
@@ -406,53 +412,61 @@ public InputStream openStream() throws IOException {
+      return s3Client.headObject(headObjectRequest);
+    } catch (NoSuchKeyException | NoSuchBucketException e) {
+      var fileNotFound = new FileNotFoundException(fileName);
+      fileNotFound.addSuppressed(e);
+      throw fileNotFound;
     }
   }
 
   /**
-   * Download S3 object and return its {@link ObjectMetadata}. Recognize the possible case of object
-   * not found (HTTP 404) by throwing more useful {@link FileNotFoundException}.
+   * Download S3 object and return its {@link GetObjectResponse}. Recognize the possible case of
+   * object not found (HTTP 404) by throwing more useful {@link FileNotFoundException}.
    *
-   * @param fileName the simple file name for the object
    * @param s3Key the S3 object key
    * @param tempDataFile where to store the downloaded object
    * @return the meta data
-   * @throws IOException eith {@link FileNotFoundException} or an AWS runtime exception
+   * @throws IOException with {@link FileNotFoundException} or an AWS runtime exception
    */
-  private ObjectMetadata downloadS3Object(String fileName, String s3Key, Path tempDataFile)
-      throws IOException {
-    try {
-      final var downloadRequest = new GetObjectRequest(s3BucketName, s3Key);
-      var metaData = s3Client.getObject(downloadRequest, tempDataFile.toFile());
-      if (metaData == null) {
-        throw new FileNotFoundException(fileName);
-      }
-      return metaData;
-    } catch (AmazonS3Exception ex) {
-      if (ex.getStatusCode() == AWS_NOT_FOUND_STATUS_CODE) {
-        throw new FileNotFoundException(convertS3KeyToFileName(s3Key));
-      } else {
-        throw ex;
-      }
-    }
+  private GetObjectResponse downloadS3Object(String s3Key, Path tempDataFile) {
+    /* Gather information needed to prepare the S3TransferManager */
+    String bucketLocation =
+        s3Client
+            .getBucketLocation(GetBucketLocationRequest.builder().bucket(s3BucketName).build())
+            .locationConstraintAsString();
+    // bucketLocation is null if the location is us-east-1, so we need to handle that result
+    // specifically
+    Region region =
+        Region.of(
+            StringUtils.isNotBlank(bucketLocation) ? bucketLocation : Region.US_EAST_1.toString());
+    S3TransferManager s3TransferManager =
+        DefaultS3TransferManager.builder()
+            .s3Client(SharedS3Utilities.createS3AsyncClient(region))
+            .build();
+    GetObjectRequest getObjectRequest =
+        GetObjectRequest.builder().bucket(s3BucketName).key(s3Key).build();
+    DownloadFileRequest downloadFileRequest =
+        DownloadFileRequest.builder()
+            .getObjectRequest(getObjectRequest)
+            .destination(tempDataFile)
+            .addTransferListener(LoggingTransferListener.create())
+            .build();
+
+    FileDownload downloadFile = s3TransferManager.downloadFile(downloadFileRequest);
+    return downloadFile.completionFuture().join().response();
   }
```

Comment:
This method previously handled AWS not found errors by throwing `FileNotFoundException`.  We need to preserve that behavior in a manner appropriate to the new API.

Catch the `CompletionException` and use an instanceof on the exception's cause to determine whether to switch to a `FileNotFound`.

Add the throws back in and change the end of the method to this:

```
    try {
      FileDownload downloadFile = s3TransferManager.downloadFile(downloadFileRequest);
      return downloadFile.completionFuture().join().response();
    } catch (CompletionException e) {
      final var cause = e.getCause();
      if (cause instanceof NoSuchKeyException || cause instanceof NoSuchBucketException) {
        final var fileName = convertS3KeyToFileName(s3Key);
        final var fileNotFound = new FileNotFoundException(fileName);
        fileNotFound.addSuppressed(e);
        throw fileNotFound;
      } else {
        throw e;
      }
    }
```

Reply:
I originally didn't include this since we were handling it in the other helper method which was already being called prior to this one (therefore the file definitely exists by the point this method is called), but I can understand your point and do agree in the end (unless the above changes your mind). I can make the change though - a defensive approach is probably best anyways.

Reply:
All set - I fully agree with you after all, let me know if this looks good.